### PR TITLE
Local SwingSet worker debugging

### PR DESCRIPTION
**Pull request by michaelfig**
These changes are to adjust some incompatibilities introduced in SwingSet that prevented debugging the sim-chain under Node.js.

- allow `$SWINGSET_WORKER_TYPE` to override SwingSet's config file default instead of providing a too-soft-to-be-useful default
- use a fake meter value instead of crashing when `SWINGSET_WORKER_TYPE=local`.

@kriskowal and I are discussing trying to support such debugging modes, starting with:

```
cd packages/cosmic-swingset
NODE_OPTIONS=--inspect-brk SWINGSET_WORKER_TYPE=local make scenario3-setup scenario3-run
```

There is still more to do to make this work nicely, but it's a start.

Next:

- propagate filenames through Endo bundles as `//#sourceURL=...` when debugging
- don't strip URL prefixes during SES stack filtering when debugging
- provide a convenient way to run in this mode using the `agoric` CLI

### CODE REVIEW

**Review by FUDCo — COMMENTED** [review]
I think this is OK.  However, it's probably worth double checking that the new config pathway doesn't provide a way for a vat empowered to create new vats to create vats that escape metering if it cannot itself escape metering.  I tried to puzzle this out myself, but I think I don't quite understand metering well enough to tell.  On that basis, I think I should let Brian have the final say on approval.

**Comment by michaelfig** [comment]
> I think this is OK. However, it's probably worth double checking that the new config pathway doesn't provide a way for a vat empowered to create new vats to create vats that escape metering if it cannot itself escape metering.

It's already the case that a vat that has access to the `vatAdmin` (vat creation power) can create as many new vats as it likes, none of which are required to be metered.  Also, it can specify any `managerType` for that new vat.  So this isn't a new exposure.

> I tried to puzzle this out myself, but I think I don't quite understand metering well enough to tell. On that basis, I think I should let Brian have the final say on approval.

@warner, can you please weigh in on this?

**Comment by warner** [comment]
I'm still thinking this through. My initial instincts are:
* It's fine to not crash a metered local vat, I was overly aggressive when I added that assert.
* I don't like having fake metering results: if a vat can't meter, and it doesn't crash, it should just return 0, not a made-up number that will confuse us into thinking it's really being metered (especially a made-up number that will diverge from reality as XS metering changes).
* We should update all workers to behave the same way: this patch touches "local" but not "subprocess-node" or "nodeworker".
* The `env` swap is *probably* ok, it was originally added to allow two CI passes to cover both default=local and default=XS, and I'm nervous about *any* use of `env` to override behavior (it always degenerates into a sort of arms race between program code and command-line arguments and env vars, q.v. Makefiles, shell scripts, etc). But I expect that you've thought through that and didn't find any other good way to accomplish the goal.

I'm not sure what to recommend on implementing zero-metering results better. I don't like the configurable made-up numbers being in `managerOptions`, and I don't like the idea of the non-meterable workers just always lying about metering results (e.g. making `supervisor-helper` unconditionally return zero). Given that for XS it's the *manager* that learns the metering results (since the XS supervisor is written in JS and thus inside the metered code), it makes more sense for the non-XS managers to be the ones to lie about metering (rather than their supervisors). Which probably means manager-helper.js is the one that should get an option to overwrite the metering results it gets from the worker, and `manager-subprocess-xsnap.js` doesn't use that option, and the other managers look at `managerOptions.metered` to decide what to provide to `makeManagerKit`. I'm not entirely thrilled with that approach (which is why I just did `assert(!managerOptions.metered)`), but it might be the best compromise.

**Comment by michaelfig** [comment]
> I'm still thinking this through. My initial instincts are:

Thanks for these thoughts.  They help greatly.
 
> I'm not sure what to recommend on implementing zero-metering results better. I don't like the configurable made-up numbers being in `managerOptions`, and I don't like the idea of the non-meterable workers just always lying about metering results (e.g. making `supervisor-helper` unconditionally return zero).

I think now that the correct behaviour is to return a `meterUsage` of `null` to indicate "no measurement was taken".  That pushes the policy for deciding what to do all the way up to the controller and/or the the vat creator.

If that seems right to you, I'll work on this PR further to implement that and the other changes you suggested.

**Comment by michaelfig** [comment]
@warner, this is R4R.  It actually does less work than the original PR and addresses the concerns you raised above.

File: packages/SwingSet/src/kernel/kernel.js
Comment (warner) [inline packages/SwingSet/src/kernel/kernel.js:456]: Could we move this block inside the `if (metering)` block now? I think that would remove the `metering !== null` clause as well as the `assert(used !== undefined)` check.

Comment (warner) [inline packages/SwingSet/src/kernel/kernel.js:456]: Could we move this block inside the `if (metering)` block now? I think that would remove the `metering !== null` clause as well as the `assert(used !== undefined)` check.

File: packages/SwingSet/test/test-device-bridge.js
Comment (warner) [inline packages/SwingSet/test/test-device-bridge.js:133]: Since we should now tolerate `metered: true` for more worker types, do we still need these config changes?

File: packages/SwingSet/src/initializeSwingset.js
Comment (warner) [inline packages/SwingSet/src/initializeSwingset.js:284]: With this priority flip, could/should we remove the similar line from packages/solo/src/init-basedir.js ?

**Review by warner — APPROVED** [review]
Ok, so the new rules are: if a static vat requests metering when creating a child vat, but either doesn't demand `managerType: 'xs-worker'` or the defaults (a combo of `env` and `config.defaultManagerType`) result in a non-xsnap worker, then the request is silently ignored, whereas previously it would cause a vat-creation time error.

That's probably fine.

r+ but consider the questions, might be some things to tidy up

File: packages/SwingSet/test/test-device-bridge.js
Comment (michaelfig) [reply packages/SwingSet/test/test-device-bridge.js:133]: This used to override any specified `SWINGSET_WORKER_TYPE=xs-worker` (by the XS test) with `local`.  Now to accomplish the same, we need to move it to `creationOptions`.

File: packages/SwingSet/src/initializeSwingset.js
Comment (michaelfig) [reply packages/SwingSet/src/initializeSwingset.js:284]: I think we still want the caller of `ag-solo init` to be able to override the default `defaultManagerType`.

**Pull request merged**
